backend/de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/payment/callout/C_PaySelectionLine.java

```diff
@@ -33,20 +49,127 @@
+
+			paySelectionBL.updateFromInvoice(psl);
+
+			final I_C_PaySelection paySelection = getPaySelection(psl);
+			final int C_BP_BankAccount_ID = paySelection.getC_BP_BankAccount_ID();
+
+			Timestamp PayDate = paySelection.getPayDate();
+			if (PayDate == null)
+			{
+				PayDate = new Timestamp(System.currentTimeMillis());
+			}
+
+			BigDecimal OpenAmt = BigDecimal.ZERO;
+			BigDecimal DiscountAmt = BigDecimal.ZERO;
+			boolean IsSOTrx = false;
+			final String sql = "SELECT currencyConvert(invoiceOpen(i.C_Invoice_ID, 0), i.C_Currency_ID,"
+					+ "ba.C_Currency_ID, i.DateInvoiced, i.C_ConversionType_ID, i.AD_Client_ID, i.AD_Org_ID),"
+					+ " paymentTermDiscount(i.GrandTotal,i.C_Currency_ID,i.C_PaymentTerm_ID,i.DateInvoiced, ?), i.IsSOTrx "
+					+ "FROM C_Invoice_v i, C_BP_BankAccount ba "
+					+ "WHERE i.C_Invoice_ID=? AND ba.C_BP_BankAccount_ID=?";	// #1..2
+			ResultSet rs = null;
+			PreparedStatement pstmt = null;
+			try
+			{
+				pstmt = DB.prepareStatement(sql, ITrx.TRXNAME_None);
+				pstmt.setTimestamp(1, PayDate);
+				pstmt.setInt(2, psl.getC_Invoice_ID());
+				pstmt.setInt(3, C_BP_BankAccount_ID);
+				rs = pstmt.executeQuery();
+				if (rs.next())
+				{
+					OpenAmt = rs.getBigDecimal(1);
+					DiscountAmt = rs.getBigDecimal(2);
+					IsSOTrx = DisplayType.toBoolean(rs.getString(3));
+				}
+			}
+			catch (final SQLException e)
+			{
+				throw new DBException(e, sql);
+			}
+			finally
+			{
+				DB.close(rs, pstmt);
+				rs = null;
+				pstmt = null;
+			}
+
+			psl.setOpenAmt(OpenAmt);
+			psl.setPayAmt(OpenAmt.subtract(DiscountAmt));
+			psl.setDiscountAmt(DiscountAmt);
+			psl.setDifferenceAmt(BigDecimal.ZERO);
+			psl.setIsSOTrx(IsSOTrx);
+		}
+	}
+
+	private I_C_PaySelection getPaySelection(@NonNull final I_C_PaySelectionLine psl)
+	{
+		return paySelectionBL.getByIdOrError(PaySelectionId.ofRepoId(psl.getC_PaySelection_ID()));
+	}
+
+	@CalloutMethod(columnNames = { I_C_PaySelectionLine.COLUMNNAME_PayAmt })
+	public void payAmt(final I_C_PaySelectionLine psl)
+	{
+		final InvoiceId invoiceId = InvoiceId.ofRepoIdOrNull(psl.getC_Invoice_ID());
+		if (invoiceId != null)
+		{
+			final BigDecimal payAmt = psl.getPayAmt();
+
+			final I_C_PaySelection paySelection = getPaySelection(psl);
+			final Timestamp payDate = paySelection.getPayDate();
+
+			final BigDecimal discountAmt = calculateDiscountAmt(invoiceBL.getById(invoiceId), payAmt, payDate);
+			psl.setDiscountAmt(discountAmt);
+
+			setDifferenceAmt(psl);
+		}
+	}
+
+	private static BigDecimal calculateDiscountAmt(final I_C_Invoice invoice, final BigDecimal payAmt, final Timestamp payDate)
+	{
+		final String trxName = InterfaceWrapperHelper.getTrxName(invoice);
+		final int currencyId = invoice.getC_Currency_ID();
+		final int paymentTermId = invoice.getC_PaymentTerm_ID();
+		final Timestamp dateInvoiced = invoice.getDateInvoiced();
+		final BigDecimal discountAmt = DB.getSQLValueBDEx(trxName,
```

Comment:
avoid direct SQL calls in callouts.
At least encapsulate this in a method of some of the BLs.

regarding trxName, IMHO u can blindly use thread inherited.